substrate/frame/contracts/uapi/src/wasm32.rs

```diff
@@ -0,0 +1,681 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::{extract_from_slice, Result, ReturnCode};
+use crate::ReturnFlags;
+use core::marker::PhantomData;
+use scale::Encode;
+
+/// Thin-wrapper around a `u32` representing a pointer for Wasm32.
+///
+/// Only for shared references.
+///
+/// # Note
+///
+/// Can only be constructed from shared reference types and encapsulates the
+/// conversion from reference to raw `u32`.
+/// Does not allow accessing the internal `u32` value.
+#[derive(Debug, Encode)]
+#[repr(transparent)]
+pub struct Ptr32<'a, T>
+where
+	T: ?Sized,
+{
+	/// The internal Wasm32 raw pointer value.
+	///
+	/// Must not be readable or directly usable by any safe Rust code.
+	_value: u32,
+	/// We handle types like these as if the associated lifetime was exclusive.
+	marker: PhantomData<fn() -> &'a T>,
+}
+
+impl<'a, T> Ptr32<'a, T>
+where
+	T: ?Sized,
+{
+	/// Creates a new Wasm32 pointer for the given raw pointer value.
+	fn new(value: u32) -> Self {
+		Self { _value: value, marker: Default::default() }
+	}
+}
+
+impl<'a, T> Ptr32<'a, [T]> {
+	/// Creates a new Wasm32 pointer from the given shared slice.
+	fn from_slice(slice: &'a [T]) -> Self {
+		Self::new(slice.as_ptr() as u32)
+	}
+}
+
+/// Thin-wrapper around a `u32` representing a pointer for Wasm32.
+///
+/// Only for exclusive references.
+///
+/// # Note
+///
+/// Can only be constructed from exclusive reference types and encapsulates the
+/// conversion from reference to raw `u32`.
+/// Does not allow accessing the internal `u32` value.
+#[derive(Debug, Encode)]
+#[repr(transparent)]
+pub struct Ptr32Mut<'a, T>
+where
+	T: ?Sized,
+{
+	/// The internal Wasm32 raw pointer value.
+	///
+	/// Must not be readable or directly usable by any safe Rust code.
+	_value: u32,
+	/// We handle types like these as if the associated lifetime was exclusive.
+	marker: PhantomData<fn() -> &'a mut T>,
+}
+
+impl<'a, T> Ptr32Mut<'a, T>
+where
+	T: ?Sized,
+{
+	/// Creates a new Wasm32 pointer for the given raw pointer value.
+	fn new(value: u32) -> Self {
+		Self { _value: value, marker: Default::default() }
+	}
+}
+
+impl<'a, T> Ptr32Mut<'a, [T]> {
+	/// Creates a new Wasm32 pointer from the given exclusive slice.
+	fn from_slice(slice: &'a mut [T]) -> Self {
+		Self::new(slice.as_ptr() as u32)
+	}
+}
+
+impl<'a, T> Ptr32Mut<'a, T>
+where
+	T: Sized,
+{
+	/// Creates a new Wasm32 pointer from the given exclusive reference.
+	fn from_ref(a_ref: &'a mut T) -> Self {
+		let a_ptr: *mut T = a_ref;
+		Self::new(a_ptr as u32)
+	}
+}
```

Comment:
Let's get rid of this noise and use pointers where necessary as in the rv32 implementation.